### PR TITLE
Fix FUIDDAO with NUID generator sequence number out of sync

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-6410

## Changes
- Maybe update the sequence number of the nuid generator on update put
- Handle initialing max sequence number for JDAO-MDAO

### CODE REVIEW

File: src/foam/util/NUIDGenerator.js [hunk 1]
```diff
@@ -78,11 +80,9 @@ foam.CLASS({
...
       javaCode: `
         // At least 2 bits sequence number
-        if ( seqNo_.get() == 0 ) {
-          synchronized ( this ) {
-            if ( seqNo_.get() == 0 ) {
-              initMaxSeqNo();
-            }
+        synchronized ( this ) {
+          if ( ! initMaxSeqNo_.get() ) {
+            initMaxSeqNo();
```
Comment (jlhughes) [inline src/foam/util/NUIDGenerator.js]: Why is this still needed?

File: src/foam/dao/FUIDDAO.js [hunk 2]
```diff
@@ -119,6 +119,12 @@ foam.CLASS({
       javaCode: `
         if ( getPropertyInfo().isDefaultValue(obj) ) {
           getPropertyInfo().set(obj, getUIDGenerator().getNext(null));
+        } else if ( getUIDGenerator() instanceof NUIDGenerator ) {
+          // NOTE: On replay, medusa node "put" its entries (with non-zero ids)
```
Comment (jlhughes) [inline src/foam/dao/FUIDDAO.js:123]: This is part of replay in general, not specific to Medusa.

Comment (lchanmann) [reply src/foam/dao/FUIDDAO.js:123]: JDAO replay doesn't call put() so didn't run through this block.

File: src/foam/util/NUIDGenerator.js (hunk 1 above)
Comment (lchanmann) [reply src/foam/util/NUIDGenerator.js]: JDAO replay doesn't call put(), think we still need to initMaxSeqNo on the first put after replay.

File: src/foam/util/NUIDGenerator.js [hunk 3]
```diff
@@ -118,6 +112,27 @@ foam.CLASS({
...
+          if ( UIDSupport.hash(id) != getHashKey() ) {
+            Loggers.logger(getX(), this).warning(getSalt(), "id:" + id, "hash not matched");
+            return;
+          }
+
+          var hex   = undoPermute(Long.toHexString(id));
+          var mid   = Integer.parseInt(hex.substring(hex.length() - 5, hex.length() - 3), 16);
+          if ( getMachineId() % 0xff == mid ) {
+            var seqNo = Integer.parseInt(hex.substring(0, hex.length() - 5), 16);
+            if ( seqNo > seqNo_.get() ) {
```
Comment (jlhughes) [inline src/foam/util/NUIDGenerator.js:133]: This require synchronized (this) {
else two test could interleave.

File: src/foam/util/NUIDGenerator.js [hunk 4]
```diff
@@ -78,11 +80,9 @@ foam.CLASS({
       `,
       javaCode: `
         // At least 2 bits sequence number
-        if ( seqNo_.get() == 0 ) {
-          synchronized ( this ) {
-            if ( seqNo_.get() == 0 ) {
-              initMaxSeqNo();
-            }
+        synchronized ( this ) {
+          if ( ! initMaxSeqNo_.get() ) {
```
Comment (jlhughes) [inline src/foam/util/NUIDGenerator.js]: Initialization will occur once so a check before the synchronized block will avoid the synchronized call during normal operation.

File: src/foam/util/NUIDGenerator.js [hunk 5]
```diff
@@ -118,6 +115,31 @@ foam.CLASS({
...
+          if ( UIDSupport.hash(id) != getHashKey() ) {
+            Loggers.logger(getX(), this).warning(getSalt(), "id:" + id, "hash not matched");
+            return;
+          }
+
+          var hex   = undoPermute(Long.toHexString(id));
+          var mid   = Integer.parseInt(hex.substring(hex.length() - 5, hex.length() - 3), 16);
+          if ( getMachineId() % 0xff == mid ) {
+            var seqNo = Integer.parseInt(hex.substring(0, hex.length() - 5), 16);
+            if ( seqNo > seqNo_.get() ) {
```
Comment (kgrgreer) [inline src/foam/util/NUIDGenerator.js:133]: All looks correct, but I think you could remove the double check and the synchronization by using the AtomicInt:  accumulateAndGet(int x, IntBinaryOperator accumulatorFunction) method.

Comment (lchanmann) [reply src/foam/util/NUIDGenerator.js:133]: Fixed in https://github.com/kgrgreer/foam3/pull/1369.

**Pull request merged**